[MRG + 1] ILP decoding

jrmyp opened the pull request:
New decoder for attelo, using Integer Linear Programming methods

Will receive new constraints later (for better results).
Uses a third-party tool (SCIP) which installation isn't yet documented.

Commit:
ENH ILP decoding

Introduces the Integer Linear Programming-based
decoding method.
Uses SCIP (third-party tool)

Commit:
CONFIG Local settings for ILP decoding

Commit:
ENH ILP decoding transitive closure constraint

Commit:
ENH ILP new constraints - acyclicity, connectivity, etc.

- acyclicity in turns
- unique head for graph
- no backward links outside of turns
  this one is enforced to maintain global acyclicity
- connectivity

kowey commented:
I'm a bit confused about this actually.  What happened to the standalone ILP decoder? Did we fuse the two commits accidentally?

jrmyp commented:
I think they have been fused... but the standalone version being the canonical one, we don't lose anything.

Commit:
DOC ILP results no longer related to MST/MSDAG + typo

moreymat commented on stac/harness/ilp.py:
```diff
@@ -0,0 +1,188 @@
+    return tuple(pair_pos.transpose())
+
+
+def zimpl_dump(dpack, tgt_dir=None, prefix='default', decoded=False):
```
I think naming conventions would favor `dump_zimpl`.

jrmyp replied:
Renamed for clarity.

moreymat commented on stac/harness/ilp.py:
```diff
@@ -0,0 +1,188 @@
+
+    Default behavior is a dump of dpack.attach and dpack.label
+    If decoded is True, dpack.prediction will be converted
+    """
```
Could I get you to use the numpy docstring conventions? Particularly useful are "sections" like Parameters and Returns, see example here: http://sphinxcontrib-napoleon.readthedocs.org/en/latest/example_numpy.html

jrmyp replied:
Sure! I added them to `ilp.py`, and will use them from now on.

moreymat commented on stac/harness/ilp.py:
```diff
@@ -0,0 +1,188 @@
+            for lis in data)
+
+def mk_zimpl_input(dpack, data_dir):
+    """ Create ZIMPL input file tuned to the current data directory """
```
Could you flesh out the docstring here? More information about parameters and result would be great, plus a brief description of what this is doing.

jrmyp replied:
Done. Hope it's clearer.

moreymat commented on stac/harness/ilp.py:
```diff
@@ -0,0 +1,188 @@
+    return input_path
+
+
+def scip_load(dpack, fn):
```
Naming convention: `load_scip` or possibly a more precise name.

jrmyp replied:
Renamed for clarity.

moreymat commented on stac/harness/local.py:
```diff
@@ -359,7 +363,7 @@ def _want_details(econf):
     has_intra_oracle = has.intra and (kids.intra.oracle or kids.inter.oracle)
     return (has_maxent and
-            ('mst' in econf.parser.key or 'astar' in econf.parser.key) and
+            any(k in econf.parser.key for k in ('mst', 'astar', 'ilp')) and
```
for k in set(...) ?

jrmyp replied:
Isn't a tuple enough?

moreymat replied:
A tuple works fine but it feels weird here. We don't want to pack together a handful of (possibly heterogeneous) values and iterate over each of them but to check against a set of possible decoders, hence the recommended use of set or rather frozenset to clearly state the intended semantics.

Commit:
MAINT+DOC Fleshed out doc, renaming for clarity

Commit:
ENH ILP decoder label prediction

ILP decoding now uses label classifier output as well as
attachment, and outputs labeled edges directly

The ZIMPL template has been updated to reflect this change

Commit:
CONFIG Bypass configuration manager for ILP decoder

Now that ILP takes care of labeling, we don't want
to run post-labeling after applying it.

Hence a new 'bypass' configuration. May be renamed.

Commit:
Merge branch 'master' into ilp-rels

jrmyp commented:
Just added the label prediction to complete the PR.

Commit:
CONFIG Ignore intra-inter for ILP

Also added frozenset for clearer semantics.

Commit:
MAINT+DOC Renaming of dump function and extra doc

moreymat commented on stac/harness/ilp.py:
```diff
@@ -0,0 +1,284 @@
+    """ Create ZIMPL input files tuned to a datapack
+
+    This creates two files:
+    ``turn.dat`` contains turn lenghts, offsets and indexes for the document
```
typo: lengths

moreymat commented on stac/harness/ilp/template.zpl:
```diff
@@ -0,0 +1,77 @@
+# No zero-prob links
+subto no_zero_att:
+    forall <i,j> in EDUs*EDUs:
+        if PATT[i, j] == 0
```
There is an implicit assumption here that scores are probabilities.
If this is the case, you should write this explicitly in `ilp.py` and add some checks where appropriate to prevent calling the ILP decoder with a non-probabilistic classifier.

Also, this line is the only use of the attachment scores in the ILP solver.
So, it seems to me that the optimization is on the labelling scores only.
If I am right, this means your decoder does not really use the attachment classifier, which is really different from what we have been doing until now, and this approach could be ported trivially to the other decoders.

In any case, what you are doing here is not clear and needs to be thoroughly documented.

moreymat commented on stac/harness/local.py:
```diff
@@ -230,10 +234,16 @@ def _core_parsers(klearner):
+        return joint + post + bypass
     else:
-        return post
+        return post + bypass
```
Are you sure about this?

jrmyp replied:
Ah, right, the ILP decoder cannot work without its scores. Fixing this.

Commit:
ENH Adding speaker info to ILP, new RFC definitions

Commit:
Merge branch 'gather-strip-mode' into ilp

Commit:
MAINT Higher time limit for decoding

Commit:
FIX No bypass (ILP) decoders if no probs are available

jrmyp commented:
@moreymat Every mentioned issue has been fixed and the ILP decoder works as it is.

I propose to merge now. I'll create a new branch for hyperparameter optimisation, which is too clunky for this PR.

moreymat commented:
Thanks, we'll see how it goes.
